### PR TITLE
Support both protobuf 3 and 4

### LINKED ISSUE #181: [Feature Request] Review options for allowing protobuf 3.x to work with this library
### Is your feature request related to a problem? Please describe.

Many libraries have not upgraded to 4.x while many have (including us). Depending on libraries that haven't is causing a problem for users. Most libraries seem to know they need to move to protobuf 4.x but just haven't done it.

### Describe the solution you'd like

Consider options for supporting both. This could likely take the form of:

* Generating proto 3.x form in addition to proto 4.x form
* Changing generation script to have the `__all__` from the `__init__` of the proto packages switch to importing the 3.x generated code if determined to be on 3.x
* Adding a CI step that confirms proto 3 works with tests and is generated properly
* Experiment with ways to have 4.x be the default depended on version but that 3.x works. This may prove to be a bit difficult (can we have an "extra" that downgrades? can we ask users to force downgrade/shade?). IIRC Poetry doesn't support transitive overrides for example.
  * I don't think we want 4.x to be an optional extra, I think we want it default because of a bug in 3.x w/ C static state sharing across imports which harms our sandbox.

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

* Changed `protobuf` dependency from `~4.21` to `>=3.19`
* Generated protos from downgraded protobuf version
* Changed CI to have one target downgrade and check gen'd code and to run tests while downgraded
* Added test for unsandboxed proto class use in workflow
* Added docs to explain some things
* Runtime check fix in sandbox

## Why?

Users need protobuf 3.x library in Python sometimes even though it is not the recommended version

## Checklist

1. Closes #181

### CODE REVIEW

File: scripts/gen_protos.py [hunk 1]
```diff
@@ -114,6 +113,31 @@ def fix_generated_output(base_path: Path):
 
 
 if __name__ == "__main__":
+    # Due to issues with the Python protobuf 3.x vs protobuf 4.x libraries, we
+    # must require that grpcio tools be on 1.48.x and protobuf be on 3.x for
+    # generation of protos. We can't check __version__ on the module (not
+    # present), and we can't use importlib.metadata due to its absence in 3.7,
+    # so we just run pip and check there.
+    proc = subprocess.run(
```
Comment (bergundy) [inline scripts/gen_protos.py:121]: nit: use `check_output` instead of run.

**Review by bergundy — APPROVED** [review]
Approved.
Just one question and one nit.

So protos generated with protobuf 3.x can be used with protobuf 4.x?
What are the implications of that?

File: scripts/gen_protos.py (hunk 1 above)
Comment (cretz) [reply scripts/gen_protos.py:121]: Ah, thanks. Will do if I revisit.

**Comment by cretz** [comment]
> So protos generated with protobuf 3.x can be used with protobuf 4.x?  What are the implications of that?

The generated code is supposed to work with both libraries. I couldn't figure out the negative implications and it scared me so I did the following:

* Explicitly guaranteed that all tests cases by default are using protobuf 4.x
* Added a env var override that allows you to run test cases with protobuf 3.x
* Created a CI task that does run the entire suite with protobuf 3.x
* Added a new test that uses a user-defined, unsandboxed protobuf file as workflow arg/result

I have seen no issues yet

**Pull request merged**